src/hobbits-core/bitarray.cpp

```diff
@@ -771,6 +771,122 @@ QSharedPointer<BitArray> BitArray::fromString(QString bitArraySpec, QStringList
     }
 }
 
+char BitArray::hexTable(qint64 nibble) const {
+    char chr;
+    if (nibble == 0) {
+        chr = '0';
+    } else if (nibble == 1) {
+        chr = '1';
+    } else if (nibble == 2) {
+        chr = '2';
+    } else if (nibble == 3) {
+        chr = '3';
+    } else if (nibble == 4) {
+        chr = '4';
+    } else if (nibble == 5) {
+        chr = '5';
+    } else if (nibble == 6) {
+        chr = '6';
+    } else if (nibble == 7) {
+        chr = '7';
+    } else if (nibble == 8) {
+        chr = '8';
+    } else if (nibble == 9) {
+        chr = '9';
+    } else if (nibble == 10) {
+        chr = 'a';
+    } else if (nibble == 11) {
+        chr = 'b';
+    } else if (nibble == 12) {
+        chr = 'c';
+    } else if (nibble == 13) {
+        chr = 'd';
+    } else if (nibble == 14) {
+        chr = 'e';
+    } else if (nibble == 15) {
+        chr = 'f';
+    } else {
+        chr = 'o';
+    }
+    return chr;
+}
+
+QString BitArray::toBin() const {
+    QString str = "";
+
+    for (quint64 i = 0; i < this->sizeInBits(); i ++) {
+        if (i < this->sizeInBits() && i >= 0) {
+            if (this->at(i) == false) {
+                str+="0";
+            } else {
+                str+="1";
+            }
+        }
+    }
+    return str;
+}
+
+QString BitArray::toBin(qint64 start, int length) const {
+    QString str = "";
+
+    for (quint64 i = start; i < start + length; i ++) {
+        if (i < this->sizeInBits() && i >= 0) {
+            if (this->at(i) == false) {
+                str+="0";
+            } else {
+                str+="1";
+            }
+        }
+    }
+    return str;
+}
+
+QString BitArray::toHex() const {
+    QString str = "";
+    quint64 nib = 0;
+
+    for (quint64 i = 0; i < this->sizeInBits(); i ++) {
+        if (i < this->sizeInBits() && i >= 0) {
+           nib = this->parseUIntValue(i, 4);
+        }
+        str += hexTable(nib);
+    }
+    return str;
+}
+
+QString BitArray::toHex(qint64 start, int length) const {
```

Comment:
both of these toHex functions could be made much simpler, safer, and probably faster by using BitArray::readBytes and QByteArray::toHex. Also, the logic shouldn't be duplicated across both of them - one could call the other.

Reply:
if you want to support nibble precision (not byte), you could probably just overshoot, and then trim the beginning and/or end of the resulting string

Reply:
> if you want to support nibble precision (not byte), you could probably just overshoot, and then trim the beginning and/or end of the resulting string

This gets a bit weird when starting on an odd nibble. In that case, you have to trim the beginning nibble as well as read an extra byte and trim half of that. Not impossible, but I wonder if it's better than this?

The hexTable() can use a switch instead of if-else-if.... which might help with speed.